### ISSUE TITLE
recurrent csum error even after successfully corrected by btrfs scrub

### ISSUE THREAD

**Issue by tribbloid** [kdave/btrfs-progs#817]
sample output:

```
liveuser@localhost-live:~$ sudo btrfs scrub status /dev/disk/by-label/Home
UUID:             a5c81116-78a5-4edb-b57c-b08e90e1391b
Scrub started:    Wed Jun 12 22:48:20 2024
Status:           running
Duration:         0:01:45
Time left:        0:00:38
ETA:              Wed Jun 12 22:50:46 2024
Total to scrub:   396.64GiB
Bytes scrubbed:   290.10GiB  (73.14%)
Rate:             2.76GiB/s
Error summary:    csum=4
  Corrected:      4
  Uncorrectable:  0
  Unverified:     0
liveuser@localhost-live:~$ sudo btrfs scrub status /dev/disk/by-label/Home
UUID:             a5c81116-78a5-4edb-b57c-b08e90e1391b
Scrub started:    Wed Jun 12 22:48:20 2024
Status:           finished
Duration:         0:02:21
Total to scrub:   396.64GiB
Rate:             2.81GiB/s
Error summary:    csum=4
  Corrected:      4
  Uncorrectable:  0
  Unverified:     0
```

So the 4 errors was listed as "corrected" but running the same command again yield the same 4 errors.

Using dmesg indicates that these error are on the same cluster:

```
[  883.471813] BTRFS info (device nvme0n1p7): scrub: started on devid 1
[  883.604772] BTRFS warning (device nvme0n1p7): tree block 58720256 mirror 1 has bad bytenr, has 67108864 want 58720256
[  883.605394] BTRFS error (device nvme0n1p7): fixed up error at logical 58720256 on dev /dev/nvme0n1p7 physical 67108864
[  883.605397] BTRFS error (device nvme0n1p7): fixed up error at logical 58720256 on dev /dev/nvme0n1p7 physical 67108864
[  883.605400] BTRFS error (device nvme0n1p7): fixed up error at logical 58720256 on dev /dev/nvme0n1p7 physical 67108864
[  883.605402] BTRFS error (device nvme0n1p7): fixed up error at logical 58720256 on dev /dev/nvme0n1p7 physical 67108864
[ 1025.166353] BTRFS info (device nvme0n1p7): scrub: finished on devid 1 with status: 0
[ 1230.506226] BTRFS info (device nvme0n1p7): scrub: started on devid 1
[ 1230.664557] BTRFS warning (device nvme0n1p7): tree block 58720256 mirror 1 has bad bytenr, has 67108864 want 58720256
[ 1230.665320] BTRFS error (device nvme0n1p7): fixed up error at logical 58720256 on dev /dev/nvme0n1p7 physical 67108864
[ 1230.665325] BTRFS error (device nvme0n1p7): fixed up error at logical 58720256 on dev /dev/nvme0n1p7 physical 67108864
[ 1230.665328] BTRFS error (device nvme0n1p7): fixed up error at logical 58720256 on dev /dev/nvme0n1p7 physical 67108864
[ 1230.665332] BTRFS error (device nvme0n1p7): fixed up error at logical 58720256 on dev /dev/nvme0n1p7 physical 67108864
[ 1371.847410] BTRFS info (device nvme0n1p7): scrub: finished on devid 1 with status: 0
```

it appears that btrfs scrub doesn't do anything in this case

**Comment by adam900710** [comment]
Kernel version please?

I tried 6.10-rc and it correctly detects and fixed it.

```
[adam@btrfs-vm ~]$ sudo btrfs scrub start -fB /mnt/btrfs/
Starting scrub on devid 1
scrub done for b069e93c-fa69-4b46-ac41-27025aafe0eb
Scrub started:    Thu Jun 13 13:13:31 2024
Status:           finished
Duration:         0:00:00
Total to scrub:   128.34MiB
Rate:             128.34MiB/s
Error summary:    verify=1
  Corrected:      1
  Uncorrectable:  0
  Unverified:     0
WARNING: errors detected during scrubbing, 1 corrected
[adam@btrfs-vm ~]$ sudo btrfs scrub start -fB /mnt/btrfs/
Starting scrub on devid 1
scrub done for b069e93c-fa69-4b46-ac41-27025aafe0eb
Scrub started:    Thu Jun 13 13:13:34 2024
Status:           finished
Duration:         0:00:00
Total to scrub:   128.34MiB
Rate:             128.34MiB/s
Error summary:    no errors found
```

**Comment by tribbloid** [comment]
fedora 40 should be using 6.8~something, let me double check

**Comment by tribbloid** [comment]
found it: Kernel: 6.8.5-301.fc40.x86_64

Also found a new recurring problem:

```
liveuser@localhost-live:~$ sudo btrfs check --repair /dev/disk/by-label/Home
enabling repair mode
WARNING:

	Do not use --repair unless you are advised to do so by a developer
	or an experienced user, and then only after having accepted that no
	fsck can successfully repair all types of filesystem corruption. E.g.
	some software or hardware bugs can fatally damage a volume.
	The operation will start in 10 seconds.
	Use Ctrl-C to stop it.
10 9 8 7 6 5 4 3 2 1
Starting repair.
Opening filesystem to check...
Checking filesystem on /dev/disk/by-label/Home
UUID: a5c81116-78a5-4edb-b57c-b08e90e1391b
[1/7] checking root items
Fixed 0 roots.
[2/7] checking extents
checksum verify failed on 58720256 wanted 0x8b416f75 found 0xf31ed09a
No device size related problem found
[3/7] checking free space cache
cache and super generation don't match, space cache will be invalidated
[4/7] checking fs roots
[5/7] checking only csums items (without verifying data)
[6/7] checking root refs
[7/7] checking quota groups skipped (not enabled on this FS)
found 421089386496 bytes used, no error found
total csum bytes: 19137532
total tree bytes: 2635005952
total fs tree bytes: 2351742976
total extent tree bytes: 256081920
btree space waste bytes: 576841045
file data blocks allocated: 486477520896
 referenced 426042204160
liveuser@localhost-live:~$ sudo btrfs check --repair /dev/disk/by-label/Home
enabling repair mode
WARNING:

	Do not use --repair unless you are advised to do so by a developer
	or an experienced user, and then only after having accepted that no
	fsck can successfully repair all types of filesystem corruption. E.g.
	some software or hardware bugs can fatally damage a volume.
	The operation will start in 10 seconds.
	Use Ctrl-C to stop it.
10 9 8 7 6 5 4 3 2 1
Starting repair.
Opening filesystem to check...
Checking filesystem on /dev/disk/by-label/Home
UUID: a5c81116-78a5-4edb-b57c-b08e90e1391b
[1/7] checking root items
Fixed 0 roots.
[2/7] checking extents
checksum verify failed on 58720256 wanted 0x3caae502 found 0x99f99bda
No device size related problem found
[3/7] checking free space cache
cache and super generation don't match, space cache will be invalidated
[4/7] checking fs roots
[5/7] checking only csums items (without verifying data)
[6/7] checking root refs
[7/7] checking quota groups skipped (not enabled on this FS)
found 421089386496 bytes used, no error found
total csum bytes: 19137532
total tree bytes: 2635005952
total fs tree bytes: 2351742976
total extent tree bytes: 256081920
btree space waste bytes: 576841045
file data blocks allocated: 486477520896
 referenced 426042204160
```

"cache and super generation don't match, space cache will be invalidated" is totally useless

**Comment by Zygo** [comment]
"space cache will be invalidated" is typically done by the kernel during the next mount.  `check --repair` does not have to do anything special in that case (other than to check the metadata of the storage where the cache is located, which is done like any other `nodatacow` file in later stages).  There's no need to check the cache contents since the kernel will wipe them out on the next mount anyway.

The message wording could be clarified.

**Comment by adam900710** [comment]
Your original report is about scrub not fixing the corruption, but why involving btrfs-check?

Anyway btrfs-progs won't repair csum errors.

Just in case, mind to run a memtest? Something weird is happening.

**Comment by tribbloid** [comment]
@adam900710 ah sorry it should be in another issue at best. Now @Zygo has explained it, I need to verify it after a reboot

**Comment by adam900710** [comment]
Nope, kernel won't really address it at mount.

The cache can only be rebuild if some write operation is done to the offending block group.

It's more recommended just wipe the cache, and go v2 cache which is safer and faster (that's why it's the default mkfs option now).

I'm more interested in why the csum mismatch happened for the tree block and why scrub doesn't repair it.

For 6.8.x, there may be the bug of kernel, but since 6.8.x is already EOL 3 weeks ago, I strongly recommend to go 6.9.5 or newer, which fixes the kernel bug that can cause some race.

**Comment by graywolf** [comment]
I think I might have the same issue.  dm-5 and dm-10 are the same physical disk,
this is across two boots.  I am running 6.12.74 kernel.  Notice that the logical
and physical are identical in both occurrences.

Any ideas how to fix this or how to debug further?

I have RAID1 profile for data and RAID1C4 for metadata.

```
[103224.991053] BTRFS error (device dm-2): fixed up error at logical 124463056945152 on dev /dev/dm-5 physical 408018681856
[103224.991350] BTRFS error (device dm-2): fixed up error at logical 124463056945152 on dev /dev/dm-5 physical 408018681856
[103224.991592] BTRFS error (device dm-2): fixed up error at logical 124463056945152 on dev /dev/dm-5 physical 408018681856
[103224.991856] BTRFS error (device dm-2): fixed up error at logical 124463056945152 on dev /dev/dm-5 physical 408018681856
[103224.992106] BTRFS error (device dm-2): fixed up error at logical 124463056945152 on dev /dev/dm-5 physical 408018681856
[103224.992108] BTRFS error (device dm-2): bdev /dev/dm-5 errs: wr 0, rd 0, flush 0, corrupt 1, gen 0
[103224.992109] BTRFS error (device dm-2): bdev /dev/dm-5 errs: wr 0, rd 0, flush 0, corrupt 2, gen 0
[103224.992110] BTRFS error (device dm-2): bdev /dev/dm-5 errs: wr 0, rd 0, flush 0, corrupt 3, gen 0
[103224.992111] BTRFS error (device dm-2): bdev /dev/dm-5 errs: wr 0, rd 0, flush 0, corrupt 4, gen 0
[103224.992112] BTRFS error (device dm-2): bdev /dev/dm-5 errs: wr 0, rd 0, flush 0, corrupt 5, gen 0
```

```
[ 1125.529032] BTRFS error (device dm-15): fixed up error at logical 124463056945152 on dev /dev/dm-10 physical 408018681856
[ 1125.529056] BTRFS error (device dm-15): fixed up error at logical 124463056945152 on dev /dev/dm-10 physical 408018681856
[ 1125.529069] BTRFS error (device dm-15): fixed up error at logical 124463056945152 on dev /dev/dm-10 physical 408018681856
[ 1125.529093] BTRFS error (device dm-15): fixed up error at logical 124463056945152 on dev /dev/dm-10 physical 408018681856
[ 1125.529107] BTRFS error (device dm-15): fixed up error at logical 124463056945152 on dev /dev/dm-10 physical 408018681856
[ 1125.529121] BTRFS error (device dm-15): bdev /dev/dm-10 errs: wr 0, rd 0, flush 0, corrupt 1, gen 0
[ 1125.529132] BTRFS error (device dm-15): bdev /dev/dm-10 errs: wr 0, rd 0, flush 0, corrupt 2, gen 0
[ 1125.529142] BTRFS error (device dm-15): bdev /dev/dm-10 errs: wr 0, rd 0, flush 0, corrupt 3, gen 0
[ 1125.529153] BTRFS error (device dm-15): bdev /dev/dm-10 errs: wr 0, rd 0, flush 0, corrupt 4, gen 0
[ 1125.529163] BTRFS error (device dm-15): bdev /dev/dm-10 errs: wr 0, rd 0, flush 0, corrupt 5, gen 0
```

Extended smart test for the drive comes back clean, and even if not, according
to https://github.com/kdave/btrfs-progs/issues/293 it should remap the sectors
after the first correction.  Did not seem to happen.

Few observations:

I *think* the error will not pop up again during this boot.  The first time I
tried it, I canceled the scrub when I saw the error and executed it again.  The
error was not detected again.  Now it is few days (and reboots) later, and it
popped up again, in the same location and file.  However full scrub takes many
hours, so testing is slow.  Is there a way to scrub just a specific logical
address?

I get these errors during scrub (same location, correctable, but the correction
does not seem to persist) across multiple drives, but always at most one
instance (well, one logical, it counts as 5 errors for some reason) per physical
drive.

In the first run, I got exactly one instance per physical device, in the second
run not every device had the error, though I deleted ~60 TB from the array in
the mean time, so it might be the case that in those suspicious physicals no
data are stored right now.

Only change I did that I can think of is that I replaced the PCIe SATA expansion
cards I am using, but this seems like extremely weird failure mode for a faulty
device (especially since I have two).

Status of the finished scrub:

```
Scrub device /dev/dm-15 (id 7) history
Scrub started:    Sun Mar  1 09:28:53 2026
Status:           finished
Duration:         13:44:57
Total to scrub:   5.07TiB
Rate:             107.48MiB/s
Error summary:    no errors found

Scrub device /dev/dm-14 (id 8) history
Scrub started:    Sun Mar  1 09:28:53 2026
Status:           finished
Duration:         11:28:31
Total to scrub:   5.05TiB
Rate:             128.11MiB/s
Error summary:    csum=5
  Corrected:      5
  Uncorrectable:  0
  Unverified:     0

Scrub device /dev/dm-13 (id 9) history
Scrub started:    Sun Mar  1 09:28:53 2026
Status:           finished
Duration:         11:10:54
Total to scrub:   5.05TiB
Rate:             131.62MiB/s
Error summary:    no errors found

Scrub device /dev/dm-12 (id 10) history
Scrub started:    Sun Mar  1 09:28:53 2026
Status:           finished
Duration:         14:50:53
Total to scrub:   5.47TiB
Rate:             107.31MiB/s
Error summary:    no errors found

Scrub device /dev/dm-11 (id 11) history
Scrub started:    Sun Mar  1 09:28:53 2026
Status:           finished
Duration:         10:42:06
Total to scrub:   4.55TiB
Rate:             123.78MiB/s
Error summary:    no errors found

Scrub device /dev/dm-10 (id 12) history
Scrub started:    Sun Mar  1 09:28:53 2026
Status:           finished
Duration:         9:29:31
Total to scrub:   4.70TiB
Rate:             144.22MiB/s
Error summary:    csum=5
  Corrected:      5
  Uncorrectable:  0
  Unverified:     0

Scrub device /dev/dm-9 (id 13) history
Scrub started:    Sun Mar  1 09:28:53 2026
Status:           finished
Duration:         12:17:15
Total to scrub:   5.73TiB
Rate:             135.92MiB/s
Error summary:    no errors found

Scrub device /dev/dm-8 (id 14) history
Scrub started:    Sun Mar  1 09:28:53 2026
Status:           finished
Duration:         22:49:40
Total to scrub:   14.16TiB
Rate:             180.66MiB/s
Error summary:    no errors found

Scrub device /dev/dm-7 (id 15) history
Scrub started:    Sun Mar  1 09:28:53 2026
Status:           finished
Duration:         4:21:23
Total to scrub:   2.69TiB
Rate:             179.81MiB/s
Error summary:    csum=5
  Corrected:      5
  Uncorrectable:  0
  Unverified:     0

Scrub device /dev/dm-6 (id 16) history
Scrub started:    Sun Mar  1 09:28:53 2026
Status:           finished
Duration:         4:32:03
Total to scrub:   2.69TiB
Rate:             172.76MiB/s
Error summary:    csum=5
  Corrected:      5
  Uncorrectable:  0
  Unverified:     0
```

```
[ 1125.529032] BTRFS error (device dm-15): fixed up error at logical 124463056945152 on dev /dev/dm-10 physical 408018681856
[ 1125.529056] BTRFS error (device dm-15): fixed up error at logical 124463056945152 on dev /dev/dm-10 physical 408018681856
[ 1125.529069] BTRFS error (device dm-15): fixed up error at logical 124463056945152 on dev /dev/dm-10 physical 408018681856
[ 1125.529093] BTRFS error (device dm-15): fixed up error at logical 124463056945152 on dev /dev/dm-10 physical 408018681856
[ 1125.529107] BTRFS error (device dm-15): fixed up error at logical 124463056945152 on dev /dev/dm-10 physical 408018681856
[ 1125.529121] BTRFS error (device dm-15): bdev /dev/dm-10 errs: wr 0, rd 0, flush 0, corrupt 1, gen 0
[ 1125.529132] BTRFS error (device dm-15): bdev /dev/dm-10 errs: wr 0, rd 0, flush 0, corrupt 2, gen 0
[ 1125.529142] BTRFS error (device dm-15): bdev /dev/dm-10 errs: wr 0, rd 0, flush 0, corrupt 3, gen 0
[ 1125.529153] BTRFS error (device dm-15): bdev /dev/dm-10 errs: wr 0, rd 0, flush 0, corrupt 4, gen 0
[ 1125.529163] BTRFS error (device dm-15): bdev /dev/dm-10 errs: wr 0, rd 0, flush 0, corrupt 5, gen 0
[ 1850.746402] BTRFS error (device dm-15): fixed up error at logical 91261710696448 on dev /dev/dm-14 physical 408004001792
[ 1850.746427] BTRFS error (device dm-15): fixed up error at logical 91261710696448 on dev /dev/dm-14 physical 408004001792
[ 1850.746440] BTRFS error (device dm-15): fixed up error at logical 91261710696448 on dev /dev/dm-14 physical 408004001792
[ 1850.746452] BTRFS error (device dm-15): fixed up error at logical 91261710696448 on dev /dev/dm-14 physical 408004001792
[ 1850.746464] BTRFS error (device dm-15): fixed up error at logical 91261710696448 on dev /dev/dm-14 physical 408004001792
[ 1850.746476] BTRFS error (device dm-15): bdev /dev/dm-14 errs: wr 0, rd 0, flush 0, corrupt 1, gen 0
[ 1850.746487] BTRFS error (device dm-15): bdev /dev/dm-14 errs: wr 0, rd 0, flush 0, corrupt 2, gen 0
[ 1850.746498] BTRFS error (device dm-15): bdev /dev/dm-14 errs: wr 0, rd 0, flush 0, corrupt 3, gen 0
[ 1850.746509] BTRFS error (device dm-15): bdev /dev/dm-14 errs: wr 0, rd 0, flush 0, corrupt 4, gen 0
[ 1850.746519] BTRFS error (device dm-15): bdev /dev/dm-14 errs: wr 0, rd 0, flush 0, corrupt 5, gen 0
[ 5115.595110] BTRFS error (device dm-15): fixed up error at logical 249086303993856 on dev /dev/dm-6 physical 1204720435200
[ 5115.595137] BTRFS error (device dm-15): fixed up error at logical 249086303993856 on dev /dev/dm-6 physical 1204720435200
[ 5115.595149] BTRFS error (device dm-15): fixed up error at logical 249086303993856 on dev /dev/dm-6 physical 1204720435200
[ 5115.595161] BTRFS error (device dm-15): fixed up error at logical 249086303993856 on dev /dev/dm-6 physical 1204720435200
[ 5115.595172] BTRFS error (device dm-15): fixed up error at logical 249086303993856 on dev /dev/dm-6 physical 1204720435200
[ 5115.595184] BTRFS error (device dm-15): bdev /dev/dm-6 errs: wr 0, rd 0, flush 0, corrupt 1, gen 0
[ 5115.595412] BTRFS error (device dm-15): bdev /dev/dm-6 errs: wr 0, rd 0, flush 0, corrupt 2, gen 0
[ 5115.595609] BTRFS error (device dm-15): bdev /dev/dm-6 errs: wr 0, rd 0, flush 0, corrupt 3, gen 0
[ 5115.595810] BTRFS error (device dm-15): bdev /dev/dm-6 errs: wr 0, rd 0, flush 0, corrupt 4, gen 0
[ 5115.595993] BTRFS error (device dm-15): bdev /dev/dm-6 errs: wr 0, rd 0, flush 0, corrupt 5, gen 0
[ 8226.747965] BTRFS error (device dm-15): fixed up error at logical 249692968124416 on dev /dev/dm-7 physical 1811384565760
[ 8226.748162] BTRFS error (device dm-15): fixed up error at logical 249692968124416 on dev /dev/dm-7 physical 1811384565760
[ 8226.748329] BTRFS error (device dm-15): fixed up error at logical 249692968124416 on dev /dev/dm-7 physical 1811384565760
[ 8226.748489] BTRFS error (device dm-15): fixed up error at logical 249692968124416 on dev /dev/dm-7 physical 1811384565760
[ 8226.748648] BTRFS error (device dm-15): fixed up error at logical 249692968124416 on dev /dev/dm-7 physical 1811384565760
[ 8226.748805] BTRFS error (device dm-15): bdev /dev/dm-7 errs: wr 0, rd 0, flush 0, corrupt 1, gen 0
[ 8226.748957] BTRFS error (device dm-15): bdev /dev/dm-7 errs: wr 0, rd 0, flush 0, corrupt 2, gen 0
[ 8226.749100] BTRFS error (device dm-15): bdev /dev/dm-7 errs: wr 0, rd 0, flush 0, corrupt 3, gen 0
[ 8226.749267] BTRFS error (device dm-15): bdev /dev/dm-7 errs: wr 0, rd 0, flush 0, corrupt 4, gen 0
[ 8226.749411] BTRFS error (device dm-15): bdev /dev/dm-7 errs: wr 0, rd 0, flush 0, corrupt 5, gen 0
```
